Feature/add alt text to several images

JanhviHarwani opened the pull request:
See [JENKINS-62424](https://issues.jenkins.io/browse/JENKINS-62424).

### Testing done

Confirmed the memory monitor text is visible from /extensionList/hudson.diagnosis.MemoryUsageMonitor/0/  and also confirmed that the graph does not render on that page

### Proposed changelog entries

- Add alternate text for images to help visually impaired Jenkins users.


### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```
Open Issue: https://issues.jenkins.io/plugins/servlet/mobile#issue/JENKINS-62424
resolved: added null text to alt attribute to img tags wherever necessary

JanhviHarwani committed:
Merge pull request #1 from jenkinsci/master

merging current changes

janhvi-simform committed:
Added alt attribute to all img tag

janhvi-simform committed:
Added null as text in alt attribute

MarkEWaite commented:
@JanhviHarwani I'm leaving this pull request open because it might address the Jira issue that is mentioned. It does not complete the pull request template. It does not provide a description of the testing that was performed to assure that the pull request works as expected. It does not include a usable changelog entry.

Could you complete those items on the pull request?

We're happy to have more contributors to Jenkins, but we need those new contributors to follow the processes that we've outlined so that we don't waste the time of maintainers as they process pull requests that do not provide the required information.

MarkEWaite commented on .github/config.yml:
```diff
@@ -20,6 +20,6 @@ firstPRMergeComment: >
   <a href="https://www.jenkins.io/participate/" target="_blank">
    <picture>
-     <img width="600" src="https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/images/jenkins-welcome.svg">
+     <img width="600" alt="jenkins welcome page" src="https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/images/jenkins-welcome.svg">
```
Jenkins is an English language proper noun.  It needs to be capitalized.
```suggestion
     <img width="600" alt="Jenkins welcome page" src="https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/images/jenkins-welcome.svg">
```

daniel-beck replied:
Needs to be localizable.

```suggestion
     <img width="600" alt="${%Jenkins welcome page}" src="https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/images/jenkins-welcome.svg">
```

Additionally I am not sure _welcome page_ adequately describes what this is.

MarkEWaite replied:
I don't think that we can localize the text in pages that are generated by GitHub Actions.

daniel-beck replied:
Oops 👍

MarkEWaite commented on core/src/main/resources/hudson/diagnosis/MemoryUsageMonitor/index.jelly:
```diff
@@ -61,7 +61,7 @@ THE SOFTWARE.
         </j:choose>
       </div>
-      <img src="graph?type=${type}&amp;width=500&amp;height=300" />
+      <img src="graph?type=${type}&amp;width=500&amp;height=300" alt="null" />
```
Since this is the memory usage monitor graph, I think that the alt text should say that.  The word "null" won't help a disabled user understand what is being shown in the image.
```suggestion
      <img src="graph?type=${type}&amp;width=500&amp;height=300" alt="memory usage monitor graph" />
```

MarkEWaite replied:
Resolved in 904cb3e4ac582523f2db3d70c1404cad5f8facb7

MarkEWaite commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -50,13 +50,13 @@ THE SOFTWARE.
            <j:if test="${showSymlinkWarning}">
              <p>
-                <img id="symlinkalert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+                <img id="symlinkalert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="null"/>
```
Since this is showing a symlink warning, I think that the alt text should mention symlink alert
```suggestion
                <img id="symlinkalert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="symlink alert"/>
```

MarkEWaite replied:
This was the wrong suggestion.  Daniel Beck made the correct suggestion in other comments.

MarkEWaite commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -50,13 +50,13 @@ THE SOFTWARE.
            <j:if test="${showTmpDirWarning}">
              <p>
-               <img id="tmpdiralert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+               <img id="tmpdiralert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="null"/>
```
```suggestion
               <img id="tmpdiralert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="temporary directory warning"/>
```

MarkEWaite commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -124,13 +124,13 @@ THE SOFTWARE.
            <j:if test="${showSymlinkWarning}">
              <p>
-                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="warning image"/>
```
I don't think that we need to use the word "image" in alt text, since the disabled user receives hints that they are seeing the alt text for an image.  A [Harvard accessibility page](https://accessibility.huit.harvard.edu/describe-content-images) advises against including the "image of" or "image" in the alt text
```suggestion
                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="symlink warning"/>
```

MarkEWaite commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -124,13 +124,13 @@ THE SOFTWARE.
            <j:if test="${showTmpDirWarning}">
              <p>
-                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="warning image"/>
```
Since it is a temporary directory warning, let's use that as alt text
```suggestion
                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="temporary directory warning"/>
```

MarkEWaite commented on core/src/main/resources/jenkins/model/Jenkins/_404.jelly:
```diff
@@ -34,7 +34,7 @@ THE SOFTWARE.
         <l:main-panel>
             <h1 style="text-align: center">
-                <img src="${imagesURL}/rage.svg" height="179" width="154"/>
+                <img src="${imagesURL}/rage.svg" height="179" width="154" alt="null"/>
```
This is the "page not found" image.
```suggestion
                <img src="${imagesURL}/rage.svg" height="179" width="154" alt="Page not found"/>
```

MarkEWaite commented on core/src/main/resources/jenkins/model/Jenkins/_404_simple.jelly:
```diff
@@ -48,7 +48,7 @@ THE SOFTWARE.
             <body style="font-family: system-ui, sans-serif">
                 <h1 style="text-align: center">
-                    <img src="${imagesURL}/rage.svg" height="179" width="154" style="vertical-align: middle"/>
+                    <img src="${imagesURL}/rage.svg" height="179" width="154" style="vertical-align: middle" alt="null"/>
```
```suggestion
                    <img src="${imagesURL}/rage.svg" height="179" width="154" style="vertical-align: middle" alt="page not found"/>
```

MarkEWaite commented on core/src/main/resources/jenkins/model/Jenkins/oops.jelly:
```diff
@@ -32,7 +32,7 @@ THE SOFTWARE.
     <l:main-panel>
         <h1 style="text-align: center">
-          <img src="${imagesURL}/rage.svg" height="179" width="154"/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
+          <img src="${imagesURL}/rage.svg" height="179" width="154" alt="null"/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
```
```suggestion
          <img src="${imagesURL}/rage.svg" height="179" width="154" alt="Oops"/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
```

daniel-beck replied:
Decorative only. Also, `null` isn't intended as a literal value, but to represent the explicitly specified empty value (`alt=""`).
```suggestion
          <img src="${imagesURL}/rage.svg" height="179" width="154" alt=""/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
```

MarkEWaite replied:
Resolved in 4d8fd426116664abdb7d45af29ba2e226821b76d

MarkEWaite commented on core/src/main/resources/lib/hudson/executors.jelly:
```diff
@@ -41,7 +41,7 @@ THE SOFTWARE.
           <j:choose>
             <j:when test="${c.offlineCause!=null and !c.connecting}">
-              (<img src="${imagesURL}/svgs/error.svg" width="16" height="16" title="${c.offlineCause}"/> ${%offline})
+              (<img src="${imagesURL}/svgs/error.svg" width="16" height="16" title="${c.offlineCause}" alt="error image"/> ${%offline})
```
```suggestion
              (<img src="${imagesURL}/svgs/error.svg" width="16" height="16" title="${c.offlineCause}" alt="Agent offline"/> ${%offline})
```

MarkEWaite commented on core/src/main/resources/lib/layout/dropdowns/custom.jelly:
```diff
@@ -31,7 +31,7 @@ THE SOFTWARE.
         <![CDATA[
             <dd:custom>
-              <img src="jenkins.svg" />
+              <img src="jenkins.svg" alt="null"/>
```
```suggestion
              <img src="jenkins.svg" alt="Jenkins logo"/>
```

MarkEWaite replied:
My recommendation is wrong here because this is an example embedded in documentation so I think that it is better to use `alt=""` than try to describe the contents of the image.

MarkEWaite reviewed (changes requested):
Thanks for the proposals.  I've suggested different text based on the context of the images.

daniel-beck reviewed (changes requested):
This PR misunderstands the recommendation in Jira. `null` isn't intended as a literal value, but to represent the explicitly specified empty value (`alt=""`).

JanhviHarwani committed:
Merge pull request #2 from jenkinsci/master

merge jenkins master branch to feature/add-null-text

janhvi-simform committed:
Merge branch 'master' of https://github.com/JanhviHarwani/jenkins into feature/add-null-text-alt

JanhviHarwani committed:
Merge pull request #3 from jenkinsci/master

merging to local branch

janhvi-simform committed:
Resolved all the PR comments

janhvi-simform committed:
Merge branch 'feature/add-null-text-alt' of https://github.com/JanhviHarwani/jenkins into feature/add-null-text-alt

janhvi-simform committed:
Updated alt image messages for dir.jelly, _404_simple.jelly and executors.jelly

JanhviHarwani commented:
I have resolved all the comments in the PR. Thank you for taking out time to look at the PR.

daniel-beck commented on .github/config.yml:
```diff
@@ -20,6 +20,6 @@ firstPRMergeComment: >
   <a href="https://www.jenkins.io/participate/" target="_blank">
    <picture>
-     <img width="600" src="https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/images/jenkins-welcome.svg">
+     <img width="600" alt=${%Jenkins welcome page} src="https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/images/jenkins-welcome.svg">
```
This applied the wrong suggestion from the comment thread.

daniel-beck replied:
(Also I'm not sure this particular alt text is good, but that's a separate problem. I encourage you to consider what the image represents on this page, and how that would be accomplished in `alt` text.)

MarkEWaite replied:
Resolved in 87d4ab46b2b736176cc328e53e70bf6169663774

daniel-beck commented on core/src/main/resources/hudson/diagnosis/MemoryUsageMonitor/index.jelly:
```diff
@@ -61,7 +61,7 @@ THE SOFTWARE.
         </j:choose>
       </div>
-      <img src="graph?type=${type}&amp;width=500&amp;height=300" />
+      <img src="graph?type=${type}&amp;width=500&amp;height=300" alt="memory usage monitor graph" />
```
```suggestion
      <img src="graph?type=${type}&amp;width=500&amp;height=300" alt="${%memory usage monitor graph}" />
```

The same applies to all `.jelly` files. (Previously mentioned in https://github.com/jenkinsci/jenkins/pull/9296#discussion_r1609285596)

MarkEWaite replied:
Fixed in 904cb3e4ac582523f2db3d70c1404cad5f8facb7

daniel-beck commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -50,13 +50,13 @@ THE SOFTWARE.
            <j:if test="${showSymlinkWarning}">
              <p>
-                <img id="symlinkalert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+                <img id="symlinkalert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="symlink alert"/>
```
Since the text in the next line exists, there is no need for an alt text here. This icon exists only to make the notice stand out.

MarkEWaite replied:
Resolved in8d861dd926cf9ee32b77b592a84b6c6bf143ef67

daniel-beck commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -50,13 +50,13 @@ THE SOFTWARE.
            <j:if test="${showTmpDirWarning}">
              <p>
-               <img id="tmpdiralert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+               <img id="tmpdiralert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="temporary directory warning"/>
```
Since the text in the next line exists, there is no need for an alt text here. This icon exists only to make the notice stand out.

MarkEWaite replied:
Resolved in 8d861dd926cf9ee32b77b592a84b6c6bf143ef67

daniel-beck commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -124,13 +124,13 @@ THE SOFTWARE.
            <j:if test="${showSymlinkWarning}">
              <p>
-                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="symlink warning"/>
```
Since the text in the next line exists, there is no need for an alt text here. This icon exists only to make the notice stand out.

MarkEWaite replied:
Resolved in 8d861dd926cf9ee32b77b592a84b6c6bf143ef67

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/oops.jelly:
```diff
@@ -32,7 +32,7 @@ THE SOFTWARE.
     <l:main-panel>
         <h1 style="text-align: center">
-          <img src="${imagesURL}/rage.svg" height="179" width="154"/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
+          <img src="${imagesURL}/rage.svg" height="179" width="154" alt=""/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
```
Inconsistent with the change in `_404_simple.jelly`, why?

(Personally I like this one more for the same reason as I mentioned in a number of other comments.)

MarkEWaite replied:
Resolved in 23d56f1cf8928bbe8388f4c0fde302779ee594e2 by using empty alt text

daniel-beck replied:
Not true, this is the only line in this PR you didn't update 😉

MarkEWaite replied:
Thanks for catching that.  This was resolved in 4d8fd426116664abdb7d45af29ba2e226821b76d

daniel-beck committed:
Merge branch 'master' into feature/add-null-text-alt

MarkEWaite committed:
Merge branch 'master' into feature/add-null-text-alt

MarkEWaite committed:
Do not attempt to localize Jenkins welcome page alt text

MarkEWaite committed:
More clearly describe the participate link

The destination page is describing the Jenkins project, not just
Jenkins core.

Addresses the comment in:
https://github.com/jenkinsci/jenkins/pull/9296#discussion_r1608916860

MarkEWaite committed:
Internationalize memory usage monitor graph

MarkEWaite committed:
Use empty alt text when immediately followed by text

As noted by Daniel Beck, it does not help the user of the screen
reader for us to show the same fundamental text message twice.  The
image in these cases is a hint for the reader of the message text that
immediately follows the image.  Cluttering the screen reader content
with a second copy of the same message is not helpful.

MarkEWaite committed:
Use empty alt text in dropdown documentation example

MarkEWaite committed:
Use empty alt text because nearby text already has the message

No need to display the "not found" message twice in locations very
near to each other.

MarkEWaite committed:
Use empty alt text - nearby text already provides details

MarkEWaite reviewed (approved):
I think this is ready to merge.  Needs a review from at least one other maintainer

MarkEWaite commented:
I attempted to display the memory usage monitor jelly change and could not find a way to reach it.  The class is used in the system information page but the graph content is loaded JavaScript into a div on the page.  The alt text in that JavaScript is "Memory usage graph" and does not appear to be internationalized.

MarkEWaite committed:
Merge branch 'master' into feature/add-null-text-alt

daniel-beck commented:
> I attempted to display the memory usage monitor jelly change and could not find a way to reach it.

A possible URL for this page is `/extensionList/hudson.diagnosis.MemoryUsageMonitor/0/` and it appears it has always been broken (since https://github.com/jenkinsci/jenkins/commit/afddd51c7c03dabe5ed14f6719abcf946371bc98), because the `heap`/`nonHeap` path elements are not present in the image URL.

MarkEWaite committed:
Merge branch 'master' into feature/add-null-text-alt

MarkEWaite commented:
> > I attempted to display the memory usage monitor jelly change and could not find a way to reach it.
> 
> A possible URL for this page is `/extensionList/hudson.diagnosis.MemoryUsageMonitor/0/` and it appears it has always been broken (since [afddd51](https://github.com/jenkinsci/jenkins/commit/afddd51c7c03dabe5ed14f6719abcf946371bc98)), because the `heap`/`nonHeap` path elements are not present in the image URL.

Thanks.  I've opened that URL and confirmed that the updated text is displayed and that the image is not displayed

MarkEWaite commented on core/src/main/resources/hudson/model/DirectoryBrowserSupport/dir.jelly:
```diff
@@ -124,13 +124,13 @@ THE SOFTWARE.
            <j:if test="${showTmpDirWarning}">
              <p>
-                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" />
+                <img id="alert" src="${imagesURL}/svgs/warning.svg" width="16" height="16" alt="temporary directory warning"/>
```
Resolved in 8d861dd926cf9ee32b77b592a84b6c6bf143ef67

A1exKH reviewed (approved):
LGTM.

krisstern committed:
Merge branch 'master' into feature/add-null-text-alt
